betty/cache.py

```diff
@@ -1,35 +1,467 @@
+
+import aiofiles.os
+from aiofiles.ospath import getmtime
+from typing_extensions import TypeVar, deprecated
 
+from betty.concurrent import AsynchronizedLock, _AsyncLock, MultiLock
 from betty.locale import Localizer
 
 
-class _Cache:
+class _HasScopes:
+    @property
+    def scopes(self) -> Sequence[str]:
+        raise NotImplementedError
+
+
+# @todo Determine which of these we must keep
+CacheItemValueT = TypeVar('CacheItemValueT')
+CacheItemValueCoT = TypeVar('CacheItemValueCoT', covariant=True)
+CacheItemValueContraT = TypeVar('CacheItemValueContraT', contravariant=True)
+
+
+class CacheItem(_HasScopes, Generic[CacheItemValueCoT]):
+    @property
+    def id(self) -> str:
+        """
+        Get this cache item's ID.
+        """
+        raise NotImplementedError
+
+    async def value(self) -> CacheItemValueCoT:
+        """
+        Get this cache item's value.
+        """
+        raise NotImplementedError
+
+    @property
+    def modified(self) -> int | float:
+        """
+        Get the time this cache item was last modified, in seconds.
+        """
+        raise NotImplementedError
+
+
+# @todo Determine which of these we must keep
+CacheItemT = TypeVar('CacheItemT', bound=CacheItem[Any])
+CacheItemCoT = TypeVar('CacheItemCoT', covariant=True)
+CacheItemContraT = TypeVar('CacheItemContraT', contravariant=True)
+
+
+class StaticCacheItem(CacheItem[CacheItemValueCoT], Generic[CacheItemValueCoT]):
+    def __init__(
+        self,
+        id: str,
+        value: CacheItemValueCoT,
+        scopes: Sequence[str],
+        modified: int | float | None = None,
+    ):
+        self._id = id
+        self._value = value
+        self._scopes = scopes
+        self._modified = modified or datetime.now().timestamp()
+
+    @property
+    def id(self) -> str:
+        return self._id
+
+    async def value(self) -> CacheItemValueCoT:
+        return self._value
+
+    @property
+    def scopes(self) -> Sequence[str]:
+        return self._scopes
+
+    @property
+    def modified(self) -> int | float:
+        return self._modified
+
+
+CacheItemValueSetter: TypeAlias = Callable[[CacheItemValueT], Awaitable[None]]
+
+
+class Cache(Generic[CacheItemCoT, CacheItemValueContraT], _HasScopes):
+    """
+    Provide a cache.
+
+    Implementations MUST be thread-safe.
+    """
+
+    def with_scope(self, scope: str) -> Self:
+        """
+        Return a new nested cache with the given scope.
+        """
+        raise NotImplementedError
+
+    async def _lock(self, cache_item_id: str) -> _AsyncLock:
+        """
+        Get a lock for the cache item with the given ID.
+        """
+        raise NotImplementedError
+
+    async def get(self, cache_item_id: str) -> CacheItemCoT | None:
+        """
+        Get the cache item with the given ID.
+        """
+        raise NotImplementedError
+
+    async def set(self, cache_item_id: str, value: CacheItemValueContraT, *, modified: int | float | None = None) -> None:
```

Comment:
If there were to return the newly set cache item, proxies can immediately cache that as well. This means that any subsequent get will hit the proxy cache rather than the underlying cache which will often incur I/O.

Reply:
Or introduce a new `.getset()` method for this in case no proxies are used?